File: PAT_Basic_Level_by_go/1051 复数乘法/1051 复数乘法.go
/*
题目：1051 复数乘法
实现原理：极坐标相乘时模长相乘、辐角相加，再用三角函数转为直角坐标输出。
*/
package main

import (
	"bufio"
	"fmt"
	"os"
	"math"
	"strconv"
	"strings"
)

// 复数乘法：结果模长 r=r1*r2，辐角相加；小于 0.01 视为 0。
func main() {
	scanner := bufio.NewScanner(os.Stdin)
	if !scanner.Scan() {
		return
	}
	f := strings.Fields(scanner.Text())
	r1, _ := strconv.ParseFloat(f[0], 64)
	p1, _ := strconv.ParseFloat(f[1], 64)
	r2, _ := strconv.ParseFloat(f[2], 64)
	p2, _ := strconv.ParseFloat(f[3], 64)
	r := r1 * r2
	A := r * math.Cos(p1+p2)
	B := r * math.Sin(p1+p2)
	if math.Abs(A) < 0.01 {
		A = 0
	}
	if math.Abs(B) < 0.01 {
		B = 0
	}
	if B < 0 {
		fmt.Printf("%.2f-%.2fi\n", A, -B)
	} else {
		fmt.Printf("%.2f+%.2fi\n", A, B)
	}
}

File: PAT_Basic_Level_by_c/1051 复数乘法/1051 复数乘法.c
/*
 * 题目：1051 复数乘法
 * 实现原理：
 *   计算两个复数的乘积。复数采用极坐标表示：R*(cos(P)+i*sin(P))
 *   复数乘法规则：(R1, P1) * (R2, P2) = (R1*R2, P1+P2)
 *   转换为直角坐标：A + Bi，其中
 *   - A = R1*R2*cos(P1+P2)
 *   - B = R1*R2*sin(P1+P2)
 *   算法步骤：
 *   1. 读取四个实数：R1, P1, R2, P2
 *   2. 计算A和B
 *   3. 处理精度问题：若|A|<0.005则置为0，同理B
 *   4. 格式化输出A+Bi（保留两位小数）
 */
#include <stdio.h>
#include <math.h>

int main() {
    double R1, P1, R2, P2;  // 两个复数的极坐标 (R, P)
    scanf("%lf %lf %lf %lf", &R1, &P1, &R2, &P2);
    double A = R1 * R2 * cos(P1 + P2);  // 乘积的实部
    double B = R1 * R2 * sin(P1 + P2);  // 乘积的虚部
    if (fabs(A) < 0.005) A = 0;  // 消除浮点误差导致的 -0.00，按四舍五入规则归零
    if (fabs(B) < 0.005) B = 0;
    printf("%.2lf%+.2lfi\n", A, B);  // %+.2lf 强制输出正负号（负号给 B，正号给 +）
    return 0;
}